res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+// * Mixxx mapping script file for the Behringer CMD Studio 4a.
+// * Author: Craig Easton
+// * Version 1.2 (Jan 2016)
+// * Forum: http://www.mixxx.org/forums/viewtopic.php?f=7&amp;t=7868
+// * Wiki: http://www.mixxx.org/wiki/doku.php/behringer_cmd_studio_4a
+// ****************************************************************************
+
+////////////////////////////////////////////////////////////////////////
+// JSHint configuration                                               //
+////////////////////////////////////////////////////////////////////////
+/* global engine                                                      */
+/* global script                                                      */
+/* global print                                                       */
+/* global midi                                                        */
+//////////////////////////////////////////////////////////////////////// 
+
+// Master function definition.
+function BehringerCMDStudio4a() {}
+
+
+// ***************************** Global Vars **********************************
+
+// Shift/mode state variables.
+BehringerCMDStudio4a.delState = {0:false,1:false,2:false,3:false};
+BehringerCMDStudio4a.scratchState = {0:false,1:false,2:false,3:false};
+
+// Button push/release state variables.
+BehringerCMDStudio4a.pitchDec = {0:false,1:false,2:false,3:false};
+BehringerCMDStudio4a.pitchInc = {0:false,1:false,2:false,3:false};
+
+
+// ************************ Initialisation stuff. *****************************
+
+BehringerCMDStudio4a.vuMeterUpdate = function (value, group, control){
+    value=(value*15)+48;
+    switch(control) {
+    case "VuMeterL":
+        midi.sendShortMsg(0xB0, 0x7E, value);
+        break;
+    case "VuMeterR":
+        midi.sendShortMsg(0xB0, 0x7F, value);
+        break;
+    }
+}
+
+BehringerCMDStudio4a.initLEDs = function () {
+    // Initialise any LEDs that are direcctly controlled by this script.
+    // Del buttons (one for each virtual deck).
+    midi.sendShortMsg(0x90, 0x2A, 0x00);
+    midi.sendShortMsg(0x91, 0x4A, 0x00);
+    midi.sendShortMsg(0x92, 0x2A, 0x00);
+    midi.sendShortMsg(0x93, 0x4A, 0x00);
+    // Scratch buttons (one for each virtual deck).
+    midi.sendShortMsg(0x90, 0x16, 0x00);
+    midi.sendShortMsg(0x91, 0x36, 0x00);
+    midi.sendShortMsg(0x92, 0x16, 0x00);
+    midi.sendShortMsg(0x93, 0x36, 0x00);    
+}
+
+BehringerCMDStudio4a.init = function () {
+    // Initialise anything that might not be in the correct state.
+    BehringerCMDStudio4a.initLEDs();
+    // Connect the VUMeters
+    engine.connectControl("[Master]","VuMeterL","BehringerCMDStudio4a.vuMeterUpdate");
+    engine.connectControl("[Master]","VuMeterR","BehringerCMDStudio4a.vuMeterUpdate");
+}
+ 
+BehringerCMDStudio4a.shutdown = function() {
+    // Leave the deck in a properly initialised state.
+    BehringerCMDStudio4a.initLEDs();
+
+    // Disconnect the VUMeters.
+    // Maybe not! This doesn't seem to work here!

```

Comment:
I do restore the LEDs to their default illumination in initLEDs(), but in init() I have to connect the VU meters to the VuMeterL and VuMeterR parameters with connectControl(). I noticed that to disconnect a control you have to pass an additional true param to the call and assumed that since the script had connected the control in init(), it was the scripts responsibility to clean up and disconnect the control in shutdown(). However, when you try to do this an error is thrown, so I guess this isn't the case. I thus commented out the disconnect (but put a reminder in the code in case I forgot why later, or if anyone else not familiar with Mixxx was wondering about this).

I didn't see anywhere in the Wiki that said you don't have to do this clean-up, and the assumption is (generally) that if you allocate a resource you should deallocate that resource when you are finished with it.